app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperState.kt

```diff
@@ -0,0 +1,29 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.wallpapers
+
+import android.graphics.Bitmap
+
+/**
+ * Represents all state related to the Wallpapers feature.
+ *
+ * @property currentWallpaper The currently selected [Wallpaper].
+ * @property currentWallpaperBitmap The bitmap used for [currentWallpaper]. Can be null to
+ * represent when the standard background should be used.
+ * @property availableWallpapers The full list of wallpapers that can be selected.
+ */
+data class WallpaperState(
+    val currentWallpaper: Wallpaper,
+    val currentWallpaperBitmap: Bitmap?,
+    val availableWallpapers: List<Wallpaper>,
+) {
+    companion object {
+        val default = WallpaperState(
+            currentWallpaper = Wallpaper.Default,
+            currentWallpaperBitmap = null,
```

Comment:
Could we not derive the bitmap from `currentWallpaper`? It seems suspicious to hold into memory the Bitmap twice if we can derive it from the other existing state.

Reply:
@MatthewTighe Yes, I agree with the trade-off you're discussing here, and in particular, that we would want to prevent having to read the bitmap from disk every time the home screen is opened. How big are these bitmaps in memory compared to the thumbnails we're already caching?

Reply:
I don't think it's feasible to have wallpapers of 10s of MB in memory or read them from disk every single time. I would think the wallpapers are just a few 100KBs on disk before rendering and we could cache that data?

Reply:
(Also, happy to talk through options offline, of course.)

Reply:
10s of MBs is a pretty large overestimation now that you mention it. I was able to find the old Turning Red assets and it looks like they're ~3.5 MB. I expect something along those lines to be towards the higher end of what we'd include in the future. I see that 200KB size for the wallpapers on disk as well. Any hypotheses around why they'd look so bloated when inspecting them through the debugger?

I'll look into `IconMemoryCache`. Would your suggestion then be to avoid the Store/Middleware pattern entirely for wallpapers? At a glance, it seems like using a cache like that would require the same hierarchy of utility classes that we've basically got now. That said, I think there's still improvements in terms of abstractions and APIs that could be made in that case.

Reply:
> 10s of MBs is a pretty large overestimation now that you mention it. I was able to find the old Turning Red assets and it looks like they're ~3.5 MB. I expect something along those lines to be towards the higher end of what we'd include in the future. I see that 200KB size for the wallpapers on disk as well. Any hypotheses around why they'd look so bloated when inspecting them through the debugger?

There's no direct relation between the size on disk and the size of a displayed image.
When displayed we need to feed information about each pixel of the screen so for a regular phone format we'd have:
- a fullscreen image 1080 x 1800px (excluding system insets) means ~2 million pixels
- ARGB.8888 - 4 bytes per pixel (if we need transparency) will mean 8MB in memory.

And then there are of course devices with higher resolutions (and probably faster NANDs also), maybe even double which will mean double the size of the bitmap in memory.

The size in memory of a bitmap can be queried with the `.byteCount` method.

Regarding passing a bitmap through the AppStore even though shallow copies are used and we'd avoid churning through memory maybe just an identifier of the image would help achieve the end goal. 
And then use maybe the `WallpaperManager` as a wrapper over all types of storage where the image may be - internet / memory / disk.
There are some fast NANDs out there and even ignoring recent flagships most devices typically reach a few hundreds MB/s so if we can ensure the images are optimized to have an as low as possible size on disk (<1MB) (there are even AI lossless optimizers and Android also supports [more efficient file formats like WebP](https://developers.google.com/speed/webp/docs/webp_lossless_alpha_study)) then I'd say loading the images from disk should be fine. We could try profiling to be sure and avoid premature optimizations.

Reply:
`IconMemoryCache` seems like just a LRU cache there to prevent new internet calls to download recent favicons and it has a 25MB limit for all in-memory bitmaps until evicting old icons to make room for the new.
The LRU cache scenario doesn't seem applicable here.
If we decide to keep the wallpapers in memory keeping just one should be enough.

Reply:
> I'll look into IconMemoryCache. Would your suggestion then be to avoid the Store/Middleware pattern entirely for wallpapers? 

@MatthewTighe Ah no, sorry, I just used it as an example for where we already cache bitmaps. It's fine to keep it in AppState as long as we're storing/caching the contents of the file, not the rendered bytes, see also @Mugurell 's explanation above.

> ...then I'd say loading the images from disk should be fine. We could try profiling to be sure and avoid premature optimizations.

@Mugurell We know based on telemetry that reading even relatively small files can be delayed on devices, and since this file can't change and the image is needed every time we open home, I do think we should cache it.

> In place of the bitmap we could cache the image asset - serialize it to a byteArray to keep in memory (<1MB) and then use a variant of [BitmapFactory.decodeByteArray()](https://developer.android.com/reference/android/graphics/BitmapFactory#decodeByteArray(byte%5B%5D,%20int,%20int,%20android.graphics.BitmapFactory.Options)) to decode the bitmap.

👍

Reply:
I'm not sure I see a real benefit to storing this data in the AppStore if a `WallpaperManager` is still required in interactions to load and decode bitmaps, especially in the case of thumbnails when it makes less sense to have them stored in memory for the lifetime of the app.
```
override fun onCreateView() {
    consumeFrom(components.appStore, viewLifecycleOwner) {
        components.wallpaperManager.decodeWallpaper(appStore.currentWallpaperBytes)
    }
}
```

Simpler:
```
override fun onCreateView() {
    components.wallpaperManager.decodeCurrentWallpaper()
}
```

I still think there's improvements to be made to class names and API surfaces, however. Maybe a `WallpaperUseCase` would be more appropriate?